Nested content elements

ausi opened the pull request:
Fixes https://github.com/contao/contao/issues/391
Alternative to https://github.com/contao/contao/pull/5814 (without the support for slots)

ausi committed:
Draft

ausi committed:
Merge branch 5.x into feature/nested-content-elements

# Conflicts:
#	core-bundle/contao/drivers/DC_Table.php
#	core-bundle/contao/templates/twig/content_element/rapper.html.twig
#	core-bundle/src/Controller/ContentElement/AbstractContentElementController.php
#	core-bundle/src/DependencyInjection/Attribute/AsContentElement.php

ausi committed:
Removed slots

ausi committed:
Merge branch '5.x' into feature/nested-content-element

ausi committed:
Remove default slot

leofeyer committed:
Merge branch '5.x' into feature/nested-content-element

leofeyer committed:
CS

leofeyer commented on core-bundle/contao/dca/tl_content.php:
```diff
@@ -1480,6 +1487,23 @@ public function disableButton($row, $href, $label, $title, $icon, $attributes)
+	 */
+	public function editChildren($row, $href, $label, $title, $icon, $attributes)
+	{
+		return System::getContainer()->get('security.helper')->isGranted(ContaoCorePermissions::USER_CAN_ACCESS_ELEMENT_TYPE, $row['type']) ? '<a href="' . $this->addToUrl($href . '&amp;id=' . $row['id']) . '" title="' . StringUtil::specialchars($title) . '"' . $attributes . '>' . Image::getHtml($icon, $label) . '</a> ' : Image::getHtml(preg_replace('/\.svg$/i', '_.svg', $icon)) . ' ';
```
We need to check if the content element supports nesting here. Any idea how to accomplish that?

ausi replied:
I guess we need to ask the fragment registry? But I’m not sure...

leofeyer replied:
It would work like this: 1f4b3daa04b11968d0b2f9f3bf39a0c9d29b449e

But if that is correct, the method should not be called `isNested()` but `supportsNesting()` or something alike, shouldn‘t it?

ausi replied:
You are right, `supportsNesting()` is a way better name for that method 👍

leofeyer committed:
Prevent circular references

leofeyer commented:
I hit the first hurdle.

<img width="817" alt="" src="https://github.com/contao/contao/assets/1192057/ef0a3ba2-737c-45a7-a90c-e5e522a17b4c">

The ID in the URL is now `1449`. It is the ID of the element I want to copy and paste inside the wrapper element.

<img width="814" alt="" src="https://github.com/contao/contao/assets/1192057/2ebb4b00-6f83-437f-9f84-3967a29133ad">

But in order to get into the wrapper element, I have to click the "edit children" link, which sets the ID to `1450`.

<img width="818" alt="" src="https://github.com/contao/contao/assets/1192057/cc64b4ca-eefe-46f6-8b7b-b1db6698f165">

And when I paste the element, it does not duplicate ID `1449` but ID `1450` (the wrapper element itself).

**Possible solutions**

1. Always use the clipboard (as we already do in "edit multiple" mode).
2. Add a third URL parameter besides `id=` and `pid=`, which stores the ID of the source element.

However, both solutions would be a behavior change. WDYT?

ausi commented:
> Always use the clipboard (as we already do in "edit multiple" mode).

To me it looks like we are using the clipboard already, because I can paste the duplicated element into a different article.

I think it needs to work the same way for this case as I also want to paste into a different “parent”.

If I navigate to a different article the URL shows only the ID of the article: `?do=article&table=tl_content&id=1`.

So the „edit child elements“ button in your case should link to `?do=article&table=tl_content&ptable=tl_content&id=1450`. The rest would then work already I think.

leofeyer commented:
I‘ll check this tomorrow. Thank you. 😊

leofeyer committed:
Merge branch '5.x' into feature/nested-content-element

leofeyer committed:
Unset `act` and `mode` in the "edit children" URL

leofeyer committed:
Use the fragment compositor to find out if a content element allows nesting

leofeyer committed:
CS

leofeyer committed:
Filter elements that are not allowed to be nested

leofeyer committed:
Copy nested content elements with children

leofeyer committed:
Replace the example controller with the content slider controller

leofeyer committed:
Fix the unit tests

leofeyer committed:
Remove the example template

leofeyer commented:
This one is read for review I think. 🥳 

You can try it by converting the content slider of the official demo to the new slider element:

<img width="858" alt="" src="https://github.com/contao/contao/assets/1192057/d7584ba3-3823-4c56-8e05-697cc54f22d2">

leofeyer committed:
Group the legacy elements in the element typ drop-down

ausi commented on core-bundle/contao/drivers/DC_Table.php:
```diff
@@ -1063,7 +1063,7 @@ protected function copyChilds($table, $insertID, $id, $parentId)
 		$db = Database::getInstance();
 
-		if (!($GLOBALS['TL_DCA'][$table]['config']['ptable'] ?? null) && Input::get('childs') && $db->fieldExists('pid', $table) && $db->fieldExists('sorting', $table))
+		if (Input::get('childs') && $db->fieldExists('pid', $table) && $db->fieldExists('sorting', $table))
```
Could it be, that instead of this change that we need to set `$GLOBALS['TL_DCA']['tl_content']['config']['ctable']` to `array('tl_content')` in *contao/dca/tl_content.php* ?

leofeyer replied:
No, I think pointing both `ptable` and `ctable` to the same table is dangerous and will cause problems.

Can you explain why you think that we need the `!($GLOBALS['TL_DCA'][$table]['config']['ptable'] ?? null)` check? It seems like an unnecessary limitation to me. Any table with a `pid` and a `sorting` column can be copied with children, regardless of whether it has a parent table or not.

ausi replied:
Looks like this caused a bug: #7422 
As we have `tl_content` as `ctable` now in https://github.com/contao/contao/blob/0b009b91b4186a32e91ca5f762df5b6a6e4ee375/core-bundle/contao/dca/tl_content.php#L37 should we revert this change?

aschempp reviewed (commented):
Things I noticed in my quick review:
 1. I'm missing the "save & go to children" button when editing the new slider element
 2. it appears a content elements can define its allowed children. I'm not sure this is a good idea, there seems to be no way to extend this. Hardcoding this would only make sense to me if there is a specific element for withing e.g. the slider, but then that element must not be available as a general content element. I think this concept needs to be re-thought, maybe something similar like picker scopes where one can add a picker to a scope but each field can define its custom scope if necessary.
 3. What is the reason the nested content elements are passed around in the `ContentElementReference`? Shouldn't the (abstract) content element controller be responsible for knowing, finding and rendering its child elements? Also keep in mind that our fragments can use differents renderers (e.g. ESI), so this must work if the slider uses ESI, or any of its children use ESI fragments).
 4. are we intentionally support nesting of front end modules? because our `RegisterFragmentsPass` allows this for both cases, and the fragment compositor will "have" nested modules then.
 5. also, why do we need the fragment compositor at all? Can't this be part of the fragment registry (where we already have a fragment config, and we can (potentially, at some point) ask our fragment e.g. if a subelement is allowed)?
 6. We agreed not to add new methods to the DCA classes, so `tl_content:copyElement()` should be a listener instead. It should also use the new button callback object to not have to generated the HTML on its own.

leofeyer committed:
Adjust the `checkAccessToElement()` method

leofeyer commented:
1. Good catch, I will add this.

2. This is optional and not meant to be extendable as far as I understood. General content elements such as the slider do not limit the types of content elements that can be nested.

3. IDK.

4. We most certainly do not.

5. IDK.

leofeyer committed:
Show the "save and edit children" button if a content element supports nesting

ausi committed:
Use Controller::getContentElement() to render nested elements

ausi committed:
Support nested elements in content_element() in Twig

ausi commented:
Switched from `{{ render() }}` to `{{ content_element() }}` in 83ffa1551f1c0e5a49fbbf0947dbe9afab592d67.

Added support for literal nested elements in c5574225d1395014951af21f1b31a5b0c863e9b7.  
You can now use something like this in Twig:

```twig
{{ content_element('slider', {
    headline: { value: 'Slider Überschrift', unit: 'h2' },
    nested_elements: [
        { type: 'text', text: 'Lorem Ipsum...' }
    ]
}) }}
```

leofeyer commented on core-bundle/contao/library/Contao/Controller.php:
```diff
@@ -514,13 +515,27 @@ public static function getArticle($varId, $blnMultiMode=false, $blnIsInsertTag=f
+			}
+
+			$contentElementReference = $intId;
+			$strColumn = $contentElementReference->getSection();
+			$intId = $contentElementReference->getContentModel();
```
```suggestion
			$strColumn = $contentElementReference->attributes['section'];
			$intId = $contentElementReference->attributes['contentModel'];
```

`ControllerReference::$attributes` is a public property, so we could use it directly and remove the getters in our `ContentElementReference` class again. WDYT?

leofeyer commented on test-case/src/ContaoTestCase.php:
```diff
@@ -133,10 +133,15 @@ protected function mockContaoFramework(array $adapters = [], array $instances =
             $framework
                 ->method('createInstance')
-                ->willReturnCallback(static fn (string $key): mixed => $instances[$key] ?? null)
+                ->willReturnCallback(static fn (string $key): mixed => ($instances[$key] ?? null)?->__invoke())
```
```suggestion
                ->willReturnCallback(
                    static function (string $key): mixed {
                        if (!isset($instances[$key])) {
                            return null;
                        }

                        if ($instances[$key] instanceof \Closure) {
                            return $instances[$key]();
                        }
                        
                        return $instances[$key];
                    }
                )
```

This is easier to understand and does not require to convert all instances to closures in lines 136–139.

leofeyer replied:
Changed in 9d625d5ab5c82c5bb1b9e24497d399abd86f19c9.

leofeyer committed:
Adjust the ContaoTestCase::mockContaoFramework() method

leofeyer committed:
CS

leofeyer committed:
Rename "nested elements" to "nested fragments"

leofeyer committed:
Use setter injection to add the nested fragments to the content element reference

leofeyer committed:
Rename `getNestedContentElements` to `getNestedFragments`

leofeyer committed:
Remove the slider element (will be re-added in a separate PR)

ausi reviewed (commented):
✅

leofeyer commented:
Thank you @ausi.
